server/backup_runner.go

```diff
@@ -70,6 +74,82 @@ func (r *BackupRunner) Run(ctx context.Context, wg *sync.WaitGroup) {
 	}
 }
 
+// TODO(dragonly): Make best effort to assure that users could recover to at least RetentionPeriodTs ago.
+// This may require pending deleting expired backup files and binlog files.
+func (r *BackupRunner) enforceRetentionPolicy(ctx context.Context) {
+	normal := api.Normal
+	instanceList, err := r.server.store.FindInstance(ctx, &api.InstanceFind{RowStatus: &normal})
+	if err != nil {
+		log.Error("Failed to find non-archived instances", zap.Error(err))
+		return
+	}
+
+	for _, instance := range instanceList {
```

Comment:
Is this purging database backups or binary logs?

If backups, we should iterate over all databases instead of instances. In other words, what if some databases in an instances don't have backup settings?

Reply:
> what if some databases in an instances don't have backup settings?

This is not a problem because `FindBackupSetting` finds backup settings but not databases.

Overall SGTM. I'll change to iterate database for general backup (for all type of databases) and iterate instance (only for MySQL binlog), which makes the code flow clear.